flow/connectors/postgres/sink.go

```diff
@@ -0,0 +1,239 @@
+	ctx context.Context,
+	qe *QRepQueryExecutor,
+	tx pgx.Tx,
+	query string,
+	args ...interface{},
+) (int, error) {
+	defer shared.RollbackTx(tx, qe.logger)
+
+	if qe.snapshot != "" {
+		_, err := tx.Exec(ctx, "SET TRANSACTION SNAPSHOT "+QuoteLiteral(qe.snapshot))
+		if err != nil {
+			qe.logger.Error("[pg_query_executor] failed to set snapshot",
+				slog.Any("error", err), slog.String("query", query))
+			err := fmt.Errorf("[pg_query_executor] failed to set snapshot: %w", err)
+			p.Close(err)
+			return 0, err
+		}
+	}
+
+	norows, err := tx.Query(ctx, query+" limit 0", args...)
+	if err != nil {
+		return 0, err
+	}
+
+	fieldDescriptions := norows.FieldDescriptions()
+	cols := make([]string, 0, len(fieldDescriptions))
+	for _, fd := range fieldDescriptions {
+		cols = append(cols, QuoteIdentifier(fd.Name))
+	}
+	p.SetSchema(cols)
+	norows.Close()
+
+	// TODO use pgx simple query arg parsing code (it's internal, need to copy)
+	for i, arg := range args {
+		var str string
+		switch arg := arg.(type) {
+		case nil:
+			str = "null"
+		case int64:
+			str = strconv.FormatInt(arg, 10)
+		case float64:
+			str = strconv.FormatFloat(arg, 'f', -1, 64)
+		case bool:
+			str = strconv.FormatBool(arg)
+		case string:
+			str = QuoteLiteral(arg)
+		case time.Time:
+			str = arg.Truncate(time.Microsecond).Format("'2006-01-02 15:04:05.999999999Z07:00:00'")
+		default:
+			return 0, fmt.Errorf("invalid arg type: %T", arg)
+		}
+		query = strings.ReplaceAll(query, fmt.Sprintf("$%d", i+1), str)
+	}
+
+	copyQuery := fmt.Sprintf("COPY (%s) TO STDOUT", query)
+	qe.logger.Info("[pg_query_executor] executing copy", slog.String("query", copyQuery))
+	ct, err := qe.conn.PgConn().CopyTo(ctx, p.PipeWriter, copyQuery)
+	if err != nil {
+		qe.logger.Info("[pg_query_executor] failed to copy",
+			slog.String("copyQuery", copyQuery), slog.Any("error", err))
+		err = fmt.Errorf("[pg_query_executor] failed to copy: %w", err)
+		p.Close(err)
+		return 0, err
+	}
+
+	qe.logger.Info("Committing transaction")
+	if err := tx.Commit(ctx); err != nil {
+		qe.logger.Error("[pg_query_executor] failed to commit transaction", slog.Any("error", err))
+		err = fmt.Errorf("[pg_query_executor] failed to commit transaction: %w", err)
+		p.Close(err)
+		return 0, err
+	}
+
+	totalRecordsFetched := ct.RowsAffected()
+	qe.logger.Info(fmt.Sprintf("[pg_query_executor] committed transaction for query '%s', rows = %d",
+		query, totalRecordsFetched))
+	return int(totalRecordsFetched), nil
+}
+
+func (p PgCopyWriter) Close(err error) {
+	p.PipeWriter.CloseWithError(err)
+}
+
+func (p PgCopyReader) GetColumnNames() []string {
+	return p.schema.schema
+}
+
+func (p PgCopyReader) CopyInto(ctx context.Context, c *PostgresConnector, tx pgx.Tx, table pgx.Identifier) (int64, error) {
+	<-p.schema.schemaLatch
+	ct, err := c.conn.PgConn().CopyFrom(
+		ctx,
+		p.PipeReader,
+		fmt.Sprintf("COPY %s (%s) FROM STDIN", table.Sanitize(), strings.Join(p.schema.schema, ",")),
```

Comment:
no chance of `BINARY` copy?

Reply:
Could do it, but I don't expect much benefit. Especially if compression is enabled. Binary has potential compatibility issues, would want to have a bool option to enable it

See https://github.com/antirez/RESP3/blob/master/spec.md where antirez outlines text protocols being fine for wire protocol. Of course if a customer has a bunch of binary blobs they want replicated we'll have trouble then. & uuid columns fall in that category

Reply:
like the idea of being an option, file an issue for that to be revisited